File: 117-populating-next-right-pointers-in-each-node-ii/solution.ts
/**
 * Definition for Node.
 * class Node {
 *     val: number
 *     left: Node | null
 *     right: Node | null
 *     next: Node | null
 *     constructor(val?: number, left?: Node, right?: Node, next?: Node) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

class TNode {
  val: number;
  left: TNode | null;
  right: TNode | null;
  next: TNode | null;
  constructor(val?: number, left?: TNode, right?: TNode, next?: TNode) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
    this.next = next === undefined ? null : next;
  }
}

function connect(root: TNode | null): TNode | null {
  if (!root) return root;
  const traversal = [{ node: root, depth: 0 }];

  let idx = 0;
  while (!!traversal[idx]) {
    const { node, depth } = traversal[idx];
    if (node.left) traversal.push({ node: node.left, depth: depth + 1 });
    if (node.right) traversal.push({ node: node.right, depth: depth + 1 });
    if (traversal[idx - 1]?.depth === depth) {
      traversal[idx - 1].node.next = node;
    }
    idx++;
  }

  return root;
}

File: 117-populating-next-right-pointers-in-each-node-ii/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    void traversal(Node* node, int lv, vector<Node*>& nodes) {
        if(node == NULL) return;
        if(nodes.size() <= lv) nodes.push_back(node);
        else {
            nodes[lv]->next = node;
            nodes[lv] = node;
        }
        node->next = NULL;
        traversal(node->left, lv + 1, nodes);
        traversal(node->right, lv + 1, nodes);
    }

    Node* connect(Node* root) {
        vector<Node*> nodes;
        traversal(root, 0, nodes);
        return root;
    }
};
